File: app/src/main/java/com/storehop/app/data/entity/PurchaseRecord.kt
package com.storehop.app.data.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "purchase_records",
    indices = [
        Index("itemId"),
        Index("storeId"),
        Index("purchasedAt"),
        Index("userId"),
        Index("householdId"),
        Index("deletedAt"),
    ],
)
data class PurchaseRecord(
    @PrimaryKey val id: String,
    val itemId: String,
    val storeId: String?,
    val purchasedAt: Long,
    val userId: String,
    val createdAt: Long,
    val updatedAt: Long,
    val deletedAt: Long?,
    @ColumnInfo(defaultValue = "1") val pendingSync: Boolean = true,
    /** v0.7.0 multi-user household access scope. See [Item.householdId]. */
    @ColumnInfo(defaultValue = "''") val householdId: String = "",
)

File: ios/Storehop/Data/Models/PurchaseRecord.swift
import Foundation
import GRDB

/// Historical audit trail of purchases. One row inserted each time the user
/// marks an item purchased at a specific store. Soft-deleted (with matching
/// `purchasedAt`) when the user undoes a purchase, so the cross-store cascade
/// can roll back precisely without affecting unrelated concurrent purchases.
struct PurchaseRecord: Codable, FetchableRecord, MutablePersistableRecord, Identifiable, Hashable, Sendable {
    static let databaseTableName = "purchase_records"

    var id: String
    var itemId: String
    /// Nullable so older Android clients (pre-Store-tracking) can sync back
    /// records they wrote without a store. Modern writes always supply one.
    var storeId: String?
    var purchasedAt: Int64
    var userId: String
    var createdAt: Int64
    var updatedAt: Int64
    var deletedAt: Int64?
    var pendingSync: Bool

    enum Columns {
        static let id          = Column(CodingKeys.id)
        static let itemId      = Column(CodingKeys.itemId)
        static let storeId     = Column(CodingKeys.storeId)
        static let purchasedAt = Column(CodingKeys.purchasedAt)
        static let userId      = Column(CodingKeys.userId)
        static let createdAt   = Column(CodingKeys.createdAt)
        static let updatedAt   = Column(CodingKeys.updatedAt)
        static let deletedAt   = Column(CodingKeys.deletedAt)
        static let pendingSync = Column(CodingKeys.pendingSync)
    }
}
